### ISSUE TITLE
[Bug]  PTODSL-vmi, vaddc/vaddcs 的sum分别恒为0xFFFFFFFF/0，且VL=1/2/4未正确支持

### ISSUE THREAD

**Issue by chenben09** [hw-native-sys/PTOAS#1416]
### Component

PTO Dialect / ODS (include/PTO/IR)

### Description

在补充 `pto.vmi.vaddc` / `pto.vmi.vaddcs` 的 PTODSL 用例时，发现以下两个问题：

1. `vaddc/vaddcs` 编译通过且 Simulator 能够完成运行，但 sum 输出错误；
2. VMI 文档声明 `VL=1/2/4` 是合法的 compact vector 长度，但对应用例无法得到正确结果，也没有在不支持时给出明确的编译期约束。

当前稳定观察到：

- `vaddc` 的 sum 输出为全 `0xFFFFFFFF`，`i32` 下表现为全 `-1`；
- `vaddcs` 的 sum 输出为全 `0`；

### Reproduction (minimal)

```markdown
关键kernel函数：
@jit_kernel(name, params)
def kernel(
    lhs_ptr: pto.ptr(dtype, "gm"),
    rhs_ptr: pto.ptr(dtype, "gm"),
    ones_ptr: pto.ptr(dtype, "gm"),
    zeros_ptr: pto.ptr(dtype, "gm"),
    out_ptr: pto.ptr(dtype, "gm"),
    carry_ptr: pto.ptr(dtype, "gm"),
):
    lhs_ub = pto.castptr(pto.i64(lhs_ub_offset), pto.ptr(dtype, "ub"))
    rhs_ub = pto.castptr(pto.i64(rhs_ub_offset), pto.ptr(dtype, "ub"))
    out_ub = pto.castptr(pto.i64(out_ub_offset), pto.ptr(dtype, "ub"))
    carry_ub = pto.castptr(pto.i64(carry_ub_offset), pto.ptr(dtype, "ub"))
    ones_ub = pto.castptr(pto.i64(ones_ub_offset), pto.ptr(dtype, "ub"))
    zeros_ub = pto.castptr(pto.i64(zeros_ub_offset), pto.ptr(dtype, "ub"))
    pto.mte_gm_ub(lhs_ptr, lhs_ub, 0, nbytes, nburst=(1, 0, 0))
    pto.mte_gm_ub(rhs_ptr, rhs_ub, 0, nbytes, nburst=(1, 0, 0))
    pto.mte_gm_ub(ones_ptr, ones_ub, 0, nbytes, nburst=(1, 0, 0))
    pto.mte_gm_ub(zeros_ptr, zeros_ub, 0, nbytes, nburst=(1, 0, 0))
    pto.set_flag("MTE2", "V", event_id=0)
    pto.wait_flag("MTE2", "V", event_id=0)
    lhs = pto.vmi.vload(lhs_ub, pto.const(0, dtype=pto.index), size=vl)
    rhs = pto.vmi.vload(rhs_ub, pto.const(0, dtype=pto.index), size=vl)
    if masked:
        if mask_kind == "group_prefix":
            exec_mask = pto.vmi.create_mask(active_lanes, size=vl, group=group)
        elif mask_kind == "prefix":
            exec_mask = pto.vmi.create_mask(active_lanes, size=vl)
        else:
            raise ValueError(f"vmi_vaddc 不支持 mask_kind={mask_kind!r}")
    else:
        exec_mask = pto.vmi.create_mask(vl, size=vl)
    result, carry_mask = pto.vmi.vaddc(lhs, rhs, exec_mask)
    # sum 必须在 vaddc 后立即写回，缩短 result 的物理寄存器生命周期，
    # 避免后续 ones/zeros vload 或 vsel 复用/覆盖 result。
    pto.vmi.vstore(result, out_ub, pto.const(0, dtype=pto.index))
    # ones/zeros 仅用于 carry mask -> dtype_max/0 的 vsel 物化。
    ones = pto.vmi.vload(ones_ub, pto.const(0, dtype=pto.index), size=vl)
    zeros = pto.vmi.vload(zeros_ub, pto.const(0, dtype=pto.index), size=vl)
    carry_vreg = pto.vmi.vsel(carry_mask, ones, zeros)
    pto.vmi.vstore(carry_vreg, carry_ub, pto.const(0, dtype=pto.index))
    pto.set_flag("V", "MTE3", event_id=0)
    pto.wait_flag("V", "MTE3", event_id=0)
    pto.mte_ub_gm(out_ub, out_ptr, nbytes, nburst=(1, 0, 0))
    pto.mte_ub_gm(carry_ub, carry_ptr, nbytes, nburst=(1, 0, 0))
```

### Expected behavior

1.vaddc/vaddcs 的 active lane sum 与逐 lane Golden 完全一致；
2.VL=1/2/4/8/64/128/256 如果是公开支持的 VMI lane count，应当能够正确编译和运行；如果 vaddc/vaddcs 实际不支持 VL=1/2/4，应在 verifier 或 lowering 阶段给出明确的 unsupported diagnostic，而不是编译通过后返回错误结果。

### Actual behavior / error logs

```text
1、vaddc 在 VL=64 时编译通过并运行完成，但 sum 全部返回 0xFFFFFFFF：
[FAIL] vmi_vaddc_ui32_vl64_g1_a64_n_unmasked backend=run |
精度校验失败 [out: out_ptr]

ACTUAL:
array([[4294967295, 4294967295, 4294967295, ...]],
      dtype=uint32)

2. 则 VL=1/2/4 的部分 UB 地址不是 32-byte aligned：
ERROR MTE: dst_addr shound be 32byte aligned!
ERROR MTE: src_addr shound be 32byte aligned!
```

### Git commit

10883a8

### Host platform

Linux (aarch64)

### Target Ascend arch (if relevant)

a5

### PTOAS build level (if relevant)

None

**Comment by TelGome** [comment]
## 1. 初始干扰：OPPROF 捕获到了 `ZerosLike`

最初的 OPPROF 中：

```text
OPPROF_20260902112516_WPGMQJADCTIKUDTH/dump/object_dump.txt
```

记录的是：

```text
ZerosLike_4_BYTES_high_performance_1
```

原因是测试框架通过以下代码创建输出 tensor：

```python
torch.zeros(shape, dtype=torch_dtype, device="npu:0")
```

这会在 DUT 之前启动 CANN 内置的 `ZerosLike` kernel，`msprof op simulator` 因而可能只捕获该前置 kernel。该 OPPROF 中的指令不能用于判断 `vmi_vmula` 的行为。

修复方式是在 simulator 模式下使用 `torch.empty`，真机模式仍保留 `torch.zeros`，避免改变 masked/preserve 用例中未写 lane 的初始化行为：

```python
if os.environ.get("PTODSL_SIMULATOR_MODE") == "1":
    out = torch.empty(shape, dtype=torch_dtype, device="npu:0")
else:
    out = torch.zeros(shape, dtype=torch_dtype, device="npu:0")
```

`sim_dsl.sh` 显式设置：

```bash
export PTODSL_SIMULATOR_MODE=1
```

修改后，OPPROF 捕获到的 kernel 变为：

```text
vmi_vmula_vmi_vmula_f32_vl1_g1_a1_n_nomask_mix_aiv
```
## 2. 最终根因：device tensor 生命周期错误

修复前 `_alloc_host_inputs` 的逻辑为：

```python
arr = data.generate(spec)
tensor = arr.to(device="npu:0")
host_inputs.append((arr, tensor.data_ptr()))
```

这里只保存了：

- CPU `arr`；
- NPU tensor 的裸 `data_ptr()`。

没有保存 NPU `tensor` 对象本身。循环进入下一次迭代后，前一个 device tensor 失去 Python 引用，其内存可以被释放并被后续输入分配复用。最终传给 kernel 的 `acc_ptr` 成为悬空指针，并读到了后续 `rhs` 的内容。

这解释了修复前的现象：

```text
V0 = rhs
V1 = lhs
V2 = rhs
```

问题不在 `RV_VMULA` 的运算，也不在 Bisheng 的寄存器选择，而在测试框架传入了生命周期已经结束的 device tensor 裸指针。

## 3. 修复

保存 host tensor、device tensor 和裸指针三者：

```python
arr = data.generate(spec)
tensor = arr.to(device="npu:0")
host_inputs.append((arr, tensor, tensor.data_ptr()))
```

构造 launch 参数时：

```python
launch_args = [ptr for _, _, ptr in host_inputs] + list(out_ptrs)
```

只要 `host_inputs` 仍在作用域中，device tensor 就保持存活，裸指针在 kernel launch 和 `torch.npu.synchronize()` 完成前始终有效。

失败证据落盘也同步适配三元组，修复了 `inputs.npz` 生成时错误解包并被 best-effort 异常处理静默吞掉的问题。

涉及文件：

```text
tools/ptodsl-test/lib/runtime.py
tools/ptodsl-test/scripts/sim_dsl.sh
```

## 4. 修复后验证

修复后重新运行同一命令，结果为：

```text
PASS: ptodsl-test :: sim/vmi_vmula/vmi_vmula_f32_vl1_g1_a1_n_nomask

精度校验成功:
mere=0
mare=0
```

最新成功运行的 OPPROF：

```text
build/lit-res/sim/cases/vmi_vmula/
  vmi_vmula_f32_vl1_g1_a1_n_nomask/
  OPPROF_20260902131158_RAKCGEWWPRRFVLMZ/
```

修复后的寄存器值正确：

```text
RV_VLDI -> V0 = 0x408567c4  // acc
RV_VLDI -> V1 = 0xc08842e3  // lhs
RV_VLDI -> V2 = 0xbf98a634  // rhs
```

执行：

```text
RV_VMULA Dtype:F32
Binary: 0x801047c1
PC:     0x10d0d110
```

写回结果：

```text
V0 = 0x4113f410 = 9.247086
```

与 FP32 native reference 以及 FP64 golden 在测试阈值内一致。

## 5. 最终结论

各层检查结果如下：

| 层级 | 结论 |
|---|---|
| 测试输入生成 | 正确 |
| PTODSL Python 前端 | 正确 |
| PTO MLIR `vmi.vmula` 操作数顺序 | 正确 |
| device-only LLVM intrinsic 参数顺序 | 正确 |
| Bisheng 生成的 `VLDI/VMULA/VSTI` 寄存器映射 | 正确 |
| CA model 对 `RV_VMULA` 的执行 | 正确 |
| 测试 runtime 输入 tensor 生命周期 | **错误，已修复** |

最终根因是测试框架只保留 device tensor 的裸指针，没有保留 tensor 所有者，造成 NPU 内存释放和复用。该问题使 `acc_ptr` 在运行时读到 `rhs` 数据，形成了“accumulator 被绑定为 rhs”的假象。

因此，当前复现不能证明 CANN 9.0.0 Bisheng 后端存在 `vmula` accumulator lowering 错误；修复测试框架的 device tensor 生命周期后，同一 kernel、同一 Bisheng 和同一 CA model 已通过精度验证。

**Comment by TelGome** [comment]
问题1：
  根因：测试 runtime 的 device tensor 提前释放
  状态：已验证修复，VL64 PASS

问题 2：VL=1/2/4 MTE 对齐错误
  根因：测试 kernel 的 UB buffer offset 未做 32-byte 对齐
  状态：本地已有对齐修改，VL1/2/4 全部 PASS

**Comment by chenben09** [comment]
vaddc/vaddcs 为什么需要 UB 32B 对齐，而其他接口在 VL=1/2/4 未对齐却正常？

同样在 VL=1/2/4 使用连续未对齐 UB 布局，其他 VMI 接口实测正常：

| 接口 | vl1/2/4 结果 |
|---|---|
| vaddc/vaddcs  | **FAIL** |
| vadds / vabs| PASS |
| vcmp / vcmps / vsel | PASS |

### 请确认

1. **MTE（`mte_gm_ub` / `mte_ub_gm`）在 UB 地址非对齐时的定义行为是什么？**
   按 32B 块粒度舍入？越界覆写相邻区域？还是未定义？
2. 若确认 **UB 32B 对齐是硬性要求**：当前该约束无任何文档。请在 VMI/PTODSL 文档中明确
   适用操作、对齐粒度。

**Comment by TelGome** [comment]
revert vaddc/vaddcs UB 32B 对齐 commit 后使用最新版测试没问题

**Comment by TelGome** [comment]
5eb87c21a (feat(vmi): legalize physical accesses and fuse stateful streams) 给 VMI 访存加了统一的对齐检查和 stateful fallback，compiler 在 codegen 阶段识别未对齐地址并自动选择 vldas/vldus、vstus/vstas stateful 非对齐访问
